Automattic/jetpack
# Google Analytics: Add support for universal analytics for eCommerce

Ready to test.

Fixes #8181 

#### Background:

* The Jetpack Google Analytics module currently only supports "legacy" (ga.js) analytics
* This PR refactors the  module in Jetpack to add support for "universal" (analytics.js) analytics and the ecommerce plugin while maintaining support for legacy analytics.
* Which is used is driven by a new field in the jetpack_wga option: `enh_ec_tracking`
* Both "modes" support IP anonymization, tracking new orders and tracking when products are added to the cart (on single or loop pages)

* A follow-on PR will expand on this PR further by adding support for these additional enhanced analytics behaviors:

enh_ec_track_remove_from_cart
enh_ec_track_prod_impression
enh_ec_track_prod_click
enh_ec_track_prod_detail_view
enh_ec_track_checkout_started

#### Changes proposed in this Pull Request:

* Note: The original `wp-google-analytics.php` file was copied to `modules/google-analytics/classes/wp-google-analytics-legacy.php` with minor editing.
* Note: The original `wp-google-analytics.php` file was then modified to either load that legacy implementation or the new universal analytics implementation.
* Note: A `classes/wp-google-analytics-options.php` file was also added to provide a consistent interface to options for both implementations, including the following new enh_ec_tracking (on/off) flag

#### Testing instructions:

* Sign up for a Google Analytics tracking code for your test site
* Install, activate and connect Jetpack and get a Jetpack Professional Plan
* Install and activate WooCommerce
* Create a product with no SKU and at least two categories
* Create another product with a SKU and at least two categories
* Replace modules/google-analytics with the files from this branch
* Install and activate the Jetpack Google Analytics Helper plugin ( https://github.com/Automattic/jetpack-google-analytics-helper )
* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: (empty)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: unchecked
```

* Visit either product page.
* View the source and 1) ensure the Google Analytics script (ga.js) is never referenced and 2) no jQuery is present to hook .single_add_to_cart_button or .add_to_cart_button
Elements.
* Note: You should see this in the page source (for legacy/classic analytics only): 

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: unchecked
```

* Visit either product page.
* View the source and 1) ensure the Google Analytics script (ga.js) is present, 2) that the Tracking ID is present, but 3) that no jQuery is present to hook .single_add_to_cart_button or .add_to_cart_button elements.

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1
Anonymize IP Addresses: CHECKED
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: unchecked
```

* Visit either product page.
* View the source and 1) ensure the Google Analytics script (ga.js) is present and that the _anonymizeIp element is present, e.g. for legacy/classic:

```
_gaq.push(['_gat._anonymizeIp']);
```

For universal:

```
ga( 'set', 'anonymizeIp', true );
```

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1
Anonymize IP Addresses: unchecked
Track Purchase Transactions: CHECKED
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: unchecked
```

* In your Google Analytics Dashboard ( https://analytics.google.com/ ), go to Admin for your site (it looks like a gear in the lower left corner)
* Under View select Ecommerce Settings
* Enable Ecommerce if you haven't already. DO NOT ENABLE ENHANCED ECOMMERCE AT THIS TIME.

<img width="1283" alt="screen shot 2017-11-20 at 12 33 54 pm" src="https://user-images.githubusercontent.com/1595739/33041195-5a356a70-cdf2-11e7-9f7f-a86391b17f18.png">

* Back on your test site, Complete the purchase of one of each product using a gateway like Check Payment.
* Return to your Google Analytics Dashboard and go to Conversions and then Ecommerce
* Set the date range to today
* Make sure you see the sale and that the details are correct (note: it can take a few minutes for the transaction to appear in GA)

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1
Anonymize IP Addresses: unchecked
Track Purchase Transactions: CHECKED
Track Add to Cart Events: CHECKED
Enable Enhanced eCommerce: unchecked
```

* On the product listing page, choose Add to Cart for one of your products
* Then, visit the other product’s product page and add that product to your cart there.
* In your Google Analytics Dashboard, go to Real-time and then Events
* Make sure you see the addition of the SKU less and the SKU product and that the details are correct (Note: for legacy/classic ecommerce the event appears as Product: Add to Cart. For enhanced ecommerce the event is stored as UX/click/Add to Cart)

<img width="1037" alt="screen shot 2017-11-20 at 12 40 44 pm" src="https://user-images.githubusercontent.com/1595739/33041206-63eb73e8-cdf2-11e7-905b-342592d51382.png">

* In your Google Analytics Dashboard, go to Admin for your site
* Under View select Ecommerce Settings
* Enable ENHANCED Ecommerce
* Now, repeat all the tests above but with *Enable Enhanced eCommerce* CHECKED

#### Proposed changelog entry for your changes:

* Google Analytics: Add support for Universal Analytics / Enhanced eCommerce

#### See also:

* https://github.com/Automattic/jetpack/pull/7817 (earlier PR that added anonymization and basic ecommerce tracking)

cc @justinshreve
